File: app/src/main/java/com/ziro/bullet/adapters/TopicsAdapter.java
package com.ziro.bullet.adapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.ziro.bullet.R;
import com.ziro.bullet.model.Menu.Category;

import java.util.ArrayList;

public class TopicsAdapter extends RecyclerView.Adapter<TopicsAdapter.ViewHolder> {

    private ArrayList<Category> topicsArrayList;
    private LayoutInflater mInflater;
    private ItemClickListener itemClickListener;
    private Context mContext;

    public TopicsAdapter(Context context, ArrayList<Category> topicsArrayList) {
        this.mInflater = LayoutInflater.from(context);
        this.topicsArrayList = topicsArrayList;
        this.mContext = context;
    }

    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = mInflater.inflate(R.layout.search_topics_new_item, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        Category item = topicsArrayList.get(position);
        holder.topic.setText(topicsArrayList.get(position).getName());

        if (item.isFavorite()) {
            holder.follow.setImageResource(R.drawable.ic_unfollow_cat);
        } else {
            holder.follow.setImageResource(R.drawable.ic_follow_cat);
        }

        holder.itemView.setOnClickListener(v -> itemClickListener.onItemTopicClick(position, topicsArrayList.get(position)));

        holder.follow.setOnClickListener(v -> {
            if (itemClickListener != null) {
//                itemClickListener.onItemTopicClick(position, topicsArrayList.get(position));
                if (item.isFavorite()) {
                    item.setFavorite(false);
                    itemClickListener.onItemTopicUnfollowed2(position, item);
                } else {
                    item.setFavorite(true);
                    itemClickListener.onItemTopicFollowed2(position, item);
                }
            }
        });

        if (position == getItemCount() - 1) {
            if (itemClickListener != null) {
                itemClickListener.isLastItem(true);
            }
        } else {
            if (itemClickListener != null) {
                itemClickListener.isLastItem(false);
            }
        }
    }

    @Override
    public int getItemCount() {
        return topicsArrayList.size();
    }


    public static class ViewHolder extends RecyclerView.ViewHolder {
        LinearLayout card;
        TextView topic;
        ImageView follow;

        ViewHolder(View itemView) {
            super(itemView);
            card = itemView.findViewById(R.id.card);
            follow = itemView.findViewById(R.id.iv_cat_image);
            topic = itemView.findViewById(R.id.tvLabel);
        }
    }

    public ArrayList<Category> getSelected() {
        ArrayList<Category> list = new ArrayList<>();
        if (topicsArrayList != null && topicsArrayList.size() > 0) {
            for (Category item : topicsArrayList) {
                if (item.isFavorite()) {
                    list.add(item);
                }
            }
        }
        return list;
    }

    public void setClickListener(ItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
    }

    public interface ItemClickListener {
        void onItemClick(View view, int position, boolean isFav);

        void isLastItem(boolean flag);

        void onItemTopicClick(int position, Category item);

        void onItemTopicFollowed2(int position, Category item);

        void onItemTopicUnfollowed2(int position, Category item);
    }
}

File: app/src/main/java/com/ziro/bullet/adapters/NewFeed/TopicsAdapter.kt
package com.ziro.bullet.adapters.NewFeed

import android.animation.ValueAnimator
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.airbnb.lottie.LottieAnimationView
import com.bumptech.glide.Glide
import com.ziro.bullet.R
import com.ziro.bullet.activities.ChannelPostActivity
import com.ziro.bullet.activities.HashTagDetailsActivity
import com.ziro.bullet.data.TYPE
import com.ziro.bullet.data.models.topics.Topics
import com.ziro.bullet.presenter.FollowUnfollowPresenter
import com.ziro.bullet.utills.Utils
import kotlinx.android.synthetic.main.item_topics.view.container
import kotlinx.android.synthetic.main.item_topics.view.image
import kotlinx.android.synthetic.main.item_topics.view.ivSelected
import kotlinx.android.synthetic.main.item_topics.view.title
import kotlinx.android.synthetic.main.item_topics_feed.view.*
import java.util.*

class TopicsAdapter(
    private val context: Context,
    private val list: ArrayList<Topics>,
) : RecyclerView.Adapter<TopicsAdapter.ViewHolder>() {

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        fun bind(context: Context, topic: Topics) {
            val presenter = FollowUnfollowPresenter(context as Activity?)

            itemView.title.text = "#"+topic.name
//            Glide.with(context)
//                .load(topic.image)
//                .centerCrop()
//                .into(itemView.image)

            if (topic.isFavorite) {
                itemView.selected.progress = 1f
            } else {
                itemView.selected.progress = 0f
            }

            itemView.setOnClickListener {
                val intent = Intent(context, ChannelPostActivity::class.java)
                //shifa test
//                val intent = Intent(context, HashTagDetailsActivity::class.java)
                intent.apply {
                    intent.putExtra("type", TYPE.TOPIC)
                    intent.putExtra("id", topic.id)
                    intent.putExtra("mContext", topic.context)
                    intent.putExtra("name", topic.name)
                    intent.putExtra("favorite", topic.isFavorite)
                }
                context.startActivity(intent)
            }

//            if (!topic.color.isEmpty())
//                itemView.container.setCardBackgroundColor(
//                        ColorStateList.valueOf(
//                                Color.parseColor(
//                                        topic.color
//                                )
//                        )
//                )

            itemView.selected.setOnClickListener {
                Utils.followAnimation(itemView.selected, 500)
                if (topic.isFavorite) {
                    topic.isFavorite = false
                    presenter.unFollowTopic(
                            topic.id,
                            absoluteAdapterPosition
                    ) { position, flag ->
                        run {
                            if (!flag)
                                topic.isFavorite = true
                        }
                    }
                } else {
                    topic.isFavorite = true
                    presenter.followTopic(
                            topic.id,
                            absoluteAdapterPosition
                    ) { position, flag ->
                        run {
                            if (!flag)
                                topic.isFavorite = false
                        }
                    }
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view =
            LayoutInflater.from(parent.context)
                .inflate(R.layout.item_topics_feed, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(context, list[position])
    }

    override fun getItemCount(): Int {
        return list.size
    }
}